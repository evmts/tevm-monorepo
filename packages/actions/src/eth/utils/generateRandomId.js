/**
 * @returns {import("@tevm/utils").Hex}
 */
export const generateRandomId = () => {
	return `0x${Array.from(crypto.getRandomValues(new Uint8Array(16)))
		.map((b) => b.toString(16).padStart(2, '0'))
		.join('')}`
}
