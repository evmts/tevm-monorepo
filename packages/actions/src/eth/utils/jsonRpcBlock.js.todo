import { bytesToHex } from "viem"

const ethjsTxToTevmTx = (tx) => {

}

/**
 * Transoforms an ethjs block to a tevm Block
 * @param {import("@ethereumjs/block").Block} block
 * @param {boolean} includeTransactions
 * @returns {Promise<import("@tevm/actions-types").Block>}
 */
const ethjsToTevmBlock = async (
  block,
  includeTransactions,
) => {
  const transactions = block.transactions.map((tx, txIndex) =>
    includeTransactions ? ethjsTxToTevmTx(tx, block, txIndex) : bytesToHex(tx.hash())
  )
  const withdrawalsAttr =
    block.header.withdrawalsRoot !== undefined
      ? {
        withdrawalsRoot: bytesToHex(block.header.withdrawalsRoot),
        withdrawals: block.toJSON().withdrawals,
      }
      : {}
  const td = await chain.getTd(block.hash(), block.header.number)
  return {
    coinbase: block.toJSON().header?.coinbase
    number: block.toJSON().header.number,
    hash: bytesToHex(block.hash()),
    parentHash: block.toJSON().header.parentHash,
    mixHash: block.toJSON().header.mixHash,
    nonce: block.toJSON().header.nonce,
    sha3Uncles: block.toJSON().header.uncleHash,
    logsBloom: block.toJSON().header.logsBloom,
    transactionsRoot: block.toJSON().header.transactionsTrie,
    stateRoot: block.toJSON().header.stateRoot,
    receiptsRoot: block.toJSON().header.receiptTrie,
    miner: block.toJSON().header.coinbase,
    difficulty: block.toJSON().header.difficulty,
    totalDifficulty: bigIntToHex(td),
    extraData: block.toJSON().header.extraData,
    size: intToHex(utf8ToBytes(JSON.stringify(block.toJSON())).byteLength),
    gasLimit: block.toJSON().header.gasLimit,
    gasUsed: block.toJSON().header.gasUsed,
    timestamp: block.toJSON().header.timestamp,
    transactions,
    uncles: block.uncleHeaders.map((uh) => bytesToHex(uh.hash())),
    baseFeePerGas: block.toJSON().header.baseFeePerGas,
    ...withdrawalsAttr,
    blobGasUsed: block.toJSON().header.blobGasUsed,
    excessBlobGas: block.toJSON().header.excessBlobGas,
    parentBeaconBlockRoot: block.toJSON().header.parentBeaconBlockRoot,
  }
}
