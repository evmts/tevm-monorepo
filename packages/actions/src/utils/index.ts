export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
