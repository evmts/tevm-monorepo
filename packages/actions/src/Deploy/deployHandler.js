import { InvalidRequestError, RevertError } from '@tevm/errors'
import { encodeDeployData } from '@tevm/utils'
import { decodeErrorResult, isHex } from '@tevm/utils'
import { callHandler } from '../Call/callHandler.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'

/**
 * Creates a tree-shakable instance of `deployHandler` for handling the deployment of contracts to TEVM.
 * This function uses `callHandler` under the hood to execute the deployment.
 *
 * Note: This is the internal logic used by higher-level APIs such as `tevmDeploy`.
 *
 * @param {import('@tevm/node').TevmNode} client - The TEVM base client instance.
 * @param {object} [options] - Optional parameters.
 * @param {boolean} [options.throwOnFail=true] - Whether to throw an error on failure.
 * @returns {import("../Deploy/DeployHandlerType.js").DeployHandler} The deploy handler function.
 * @throws {import('./TevmCallError.js').TevmCallError} If `throwOnFail` is true, returns `TevmCallError` as value.
 *
 * @example
 * ```typescript
 * import { createTevmNode } from 'tevm/node'
 * import { deployHandler } from 'tevm/actions'
 *
 * const client = createTevmNode()
 *
 * const deploy = deployHandler(client)
 *
 * const res = await deploy({
 *   bytecode: '0x...', // Contract bytecode
 *   abi: [{...}], // ABI array
 *   args: [1, 2, 3], // Constructor arguments
 *   createTransaction: true,
 * })
 * ```
 */
export const deployHandler =
	(client, { throwOnFail: throwOnFailDefault = true } = {}) =>
	async (params) => {
		client.logger.debug(params, 'deployHandler called with params')
		// TODO Moving fast atm we should ideally add validators to zod
		let deployData
		try {
			deployData = encodeDeployData(
				/** @type {any} */ ({
					abi: params.abi,
					bytecode: params.bytecode,
					args: params.args,
				}),
			)
		} catch (e) {
			client.logger.debug(e, 'deployHandler: Unable to encode the abi, functionName, and args into hex data')
			const cause = /** @type {Error}*/ (e)
			const err = new InvalidRequestError(cause.message, { cause })

			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
				/**
				 * @type {import('@tevm/utils').Hex}
				 */
				rawData: '0x',
				executionGasUsed: 0n,
				errors: [err],
			})
		}

		client.logger.debug(deployData, 'deployHandler: Encoded abi bytecode and args into hex data to execute call')

		const result = await callHandler(client, {
			throwOnFail: throwOnFailDefault,
		})({
			...params,
			// Unlike most calls deployments default to true
			createTransaction: params.createTransaction !== undefined ? params.createTransaction : true,
			data: deployData,
			throwOnFail: false,
		})

		if (result.errors && result.errors.length > 0) {
			result.errors = result.errors.map((err) => {
				if (isHex(err.message) && err instanceof RevertError) {
					try {
						client.logger.debug(err, 'contractHandler: Contract revert error. Decoding the error')
						const decodedError = decodeErrorResult(
							/** @type {any} */ ({
								abi: params.abi,
								data: err.message,
								functionName: 'constructor',
							}),
						)
						const message = `Revert: ${decodedError.errorName} ${JSON.stringify(decodedError)}`
						client.logger.debug(message, 'Revert message decoded')
						return new RevertError(message, { cause: /** @type any*/ (err) })
					} catch (e) {
						client.logger.warn(e, 'Unable to decode revert data')
						return err
					}
				}
				return err
			})
			client.logger.debug(result.errors, 'contractHandler: Execution errors')
			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, result)
		}

		return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, result)
	}
