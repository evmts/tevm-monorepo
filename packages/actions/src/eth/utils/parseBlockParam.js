import { InvalidBlockError } from '@tevm/errors'
import { hexToBigInt, hexToBytes } from '@tevm/utils'

/**
 * @param {import('@tevm/blockchain').Chain} blockchain
 * @param {import('../../common/BlockParam.js').BlockParam} blockParam
 * @returns {Promise<bigint >}
 */
export const parseBlockParam = async (blockchain, blockParam) => {
	if (typeof blockParam === 'number') {
		return BigInt(blockParam)
	}
	if (typeof blockParam === 'bigint') {
		return blockParam
	}
	if (typeof blockParam === 'string' && blockParam.startsWith('0x')) {
		if (blockParam.length === 66) {
			const block = await blockchain.getBlock(hexToBytes(/** @type {import('@tevm/utils').Hex}*/ (blockParam)))
			return BigInt(block.header.number)
		}
		return hexToBigInt(/** @type {import('@tevm/utils').Hex}*/ (blockParam))
	}
	if (blockParam === 'safe') {
		const safeBlock = blockchain.blocksByTag.get('safe')
		// let's handle it here in case we forget to update it later
		if (safeBlock) {
			return safeBlock.header.number
		}
		throw new InvalidBlockError('safe not currently supported as block tag')
	}
	if (blockParam === 'latest' || blockParam === undefined) {
		const safeBlock = blockchain.blocksByTag.get('latest')
		// let's handle it here in case we forget to update it later
		if (safeBlock) {
			return safeBlock.header.number
		}
		throw new InvalidBlockError('latest block does not exist on chain')
	}
	if (blockParam === 'pending') {
		// for pending we need to mine a new block and then handle it
		// let's skip this functionality for now
		throw new InvalidBlockError(
			'Pending not yet supported but will be in future. Consider opening an issue or reaching out on telegram if you need this feature to expediate its release',
		)
	}
	if (blockParam === 'earliest') {
		return BigInt(1)
	}
	if (blockParam === 'finalized') {
		throw new InvalidBlockError('finalized noet yet supported for this feature')
	}
	blockchain.logger.error({ blockParam }, 'Unknown block param pased to blockNumberHandler')
	throw new InvalidBlockError(`Unknown block param ${blockParam} pased to blockNumberHandler`)
}
