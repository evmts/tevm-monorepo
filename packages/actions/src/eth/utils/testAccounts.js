import { mnemonicToAccount } from '@tevm/utils'

const oneToTen = Array.from(Array(10).keys())

export const mnemonic =
	'test test test test test test test test test test test junk'

export const testAccounts =
	/**
	 * @type {[import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount,import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount,import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount,import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount,import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount]}
	 */
	(oneToTen.map((i) => mnemonicToAccount(mnemonic, { addressIndex: i })))
