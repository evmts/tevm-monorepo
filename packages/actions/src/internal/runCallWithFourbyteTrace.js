import { bytesToHex } from '@tevm/utils'

/**
 * @typedef {`${import('@tevm/utils').Hex}-${number}`} SelectorAndSizeKey
 * @typedef {`${import('@tevm/utils').Hex}`} SelectorKey
 */
/**
 * @internal
 * Prepares a trace to collect 4-byte function selectors from contract calls
 * @param {import('@tevm/vm').Vm} vm
 * @param {import('@tevm/node').TevmNode['logger']} logger
 * @param {import('@tevm/evm').EvmRunCallOpts} params
 * @param {boolean} [lazilyRun]
 * @returns {Promise<import('@tevm/evm').EvmResult & {trace: import('../common/FourbyteTraceResult.js').FourbyteTraceResult}>}
 * @throws {never}
 */
export const runCallWithFourbyteTrace = async (vm, logger, params, lazilyRun = false) => {
	/**
	 * Map of selector-calldata_size keys to their call counts
	 * Format: "0x{selector}-{calldata_size}" -> count
	 * @type {Record<SelectorAndSizeKey, number>}
	 */
	const selectors = {}
	/**
	 * Map of selector keys to an array of calldata
	 * Format: "0x{selector}" -> [calldata1, calldata2, ...]
	 * @type {Record<SelectorKey, import('@tevm/utils').Hex[]>}
	 */
	const calldatas = {}

	/**
	 * Extract 4-byte selector and calldata size from call data
	 * @param {Uint8Array} data - Call data
	 * @returns {{key: SelectorAndSizeKey, selector: SelectorKey, calldata: import('@tevm/utils').Hex, calldataSize: number} | null} - Formatted key and components or null if data is too short
	 */
	const extractSelectorAndSize = (data) => {
		if (data.length >= 4) {
			const selector = bytesToHex(data.slice(0, 4))
			const calldata = data.slice(4)
			const calldataSize = calldata.length // Size excluding the 4-byte selector
			const key = /** @type {SelectorAndSizeKey} */ (`${selector}-${calldataSize}`)
			return { key, selector, calldata: bytesToHex(calldata), calldataSize }
		}
		return null
	}

	/**
	 * Before each call/create - collect function selectors
	 *
	 * This will capture inner-contract calls if they generate a "CALL" opcode (i.e. `this.function()`);
	 * which is not the case for a "JUMP" opcode (i.e. `function()`)
	 */
	vm.evm.events?.on('beforeMessage', async (message, next) => {
		logger.debug(message, 'runCallWithFourbyteTrace: beforeMessage event')

		// Only process CALL, DELEGATECALL, and STATICCALL operations
		// Skip CREATE and CREATE2 as they don't have function selectors
		if (message.to && message.data && message.data.length >= 4) {
			const result = extractSelectorAndSize(message.data)
			if (result) {
				// Increment counter for this selector-calldata_size combination
				selectors[result.key] = (selectors[result.key] ?? 0) + 1
				// Add this calldata to the array of calldatas for this selector
				calldatas[result.selector] = [...(calldatas[result.selector] ?? []), result.calldata]

				logger.debug(
					{
						key: result.key,
						selector: result.selector,
						calldata: result.calldata,
						calldataSize: result.calldataSize,
						count: selectors[result.key],
					},
					'runCallWithFourbyteTrace: collected selector with calldata size',
				)
			}
		}

		next?.()
	})

	if (lazilyRun) {
		// Return object with trace without running EVM
		return /** @type {any} */ ({ trace: selectors })
	}

	// Execute the call
	const runCallResult = await vm.evm.runCall(params)

	logger.debug(runCallResult, 'runCallWithFourbyteTrace: evm run call complete')
	logger.debug(selectors, 'runCallWithFourbyteTrace: collected selectors')

	return {
		...runCallResult,
		trace: {
			...selectors,
			...calldatas,
		},
	}
}
