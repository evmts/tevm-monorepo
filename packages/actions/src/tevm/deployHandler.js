import { callHandler } from './callHandler.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { encodeDeployData } from '@tevm/utils'
import {
  decodeErrorResult,
  isHex,
} from '@tevm/utils'

// TODO let's add warnings to other calls such that if the contract is still in mempool we throw a warning telling user they need to mine a block

/**
 * Creates an DeployHandler for handling deploying a contract to tevm
 * @param {import('@tevm/base-client').BaseClient} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import("@tevm/actions-types").DeployHandler}
 */
export const deployHandler =
  (client, { throwOnFail: throwOnFailDefault = true } = {}) =>
    async (params) => {
      client.logger.debug(params, 'deployHandler called with params')
      // TODO Moving fast atm we should ideally add validators to @tevm/zod
      let deployData
      try {
        deployData = encodeDeployData(
				/** @type {any} */({
            abi: params.abi,
            bytecode: params.bytecode,
            args: params.args,
          }),
        )
      } catch (e) {
        client.logger.debug(
          e,
          'deployHandler: Unable to encode the abi, functionName, and args into hex data',
        )
        /**
         * @type {import('@tevm/errors').InvalidRequestError}
         */
        const err = {
          name: 'InvalidRequestError',
          _tag: 'InvalidRequestError',
          message: /** @type {Error}*/ (e).message,
        }
        return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
          /**
           * @type {import('@tevm/utils').Hex}
           */
          rawData: '0x',
          executionGasUsed: 0n,
          errors: [err],
        })
      }

      client.logger.debug(
        deployData,
        'deployHandler: Encoded abi bytecode and args into hex data to execute call',
      )

      const result = await callHandler(client, {
        throwOnFail: throwOnFailDefault,
      })({
        ...params,
        // Unlike most calls deployments default to true
        createTransaction: params.createTransaction !== undefined ? params.createTransaction : true,
        data: deployData,
        throwOnFail: false,
      })

      if (result.errors && result.errors.length > 0) {
        result.errors = result.errors.map((err) => {
          if (isHex(err.message) && err._tag === 'revert') {
            client.logger.debug(
              err,
              'contractHandler: Contract revert error. Decoding the error',
            )
            const decodedError = decodeErrorResult(
						/** @type {any} */({
                abi: params.abi,
                data: err.message,
                functionName: 'constructor',
              }),
            )
            const message = `Revert: ${decodedError.errorName} ${JSON.stringify(
              decodedError,
            )}`
            client.logger.debug(message, 'Revert message decoded')
            return {
              ...err,
              message,
            }
          }
          return err
        })
        client.logger.debug(result.errors, 'contractHandler: Execution errors')
        return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, result)
      }

      return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, result)
    }
