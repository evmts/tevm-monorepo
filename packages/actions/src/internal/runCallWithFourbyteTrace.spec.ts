import { Address, createAddress } from '@tevm/address'
import { AdvancedContract, ErrorContract } from '@tevm/contract'
import { createTevmNode } from '@tevm/node'
import { PREFUNDED_ACCOUNTS, encodeDeployData, encodeFunctionData, hexToBytes } from '@tevm/utils'
import { toFunctionSelector } from 'viem'
import { assert, beforeEach, describe, expect, it } from 'vitest'
import { deployHandler } from '../Deploy/deployHandler.js'
import { runCallWithFourbyteTrace } from './runCallWithFourbyteTrace.js'

describe('runCallWithFourbyteTrace', () => {
	let client: ReturnType<typeof createTevmNode>
	let advancedContractAddress: Address
	let errorContractAddress: Address

	// Inner function calls with `this`
	const setAllValuesExpectedTrace = {
		[`${toFunctionSelector('function setAllValues(uint256,bool,string,address)')}-192`]: 1, // each 32 bytes: (newNumber, newBool, newString (offset), newAddress, newString (length), newString (value padded to 32 bytes))
		[`${toFunctionSelector('function setNumber(uint256)')}-32`]: 1, // each 32 bytes: (newNumber)
		[`${toFunctionSelector('function setBool(bool)')}-32`]: 1, // each 32 bytes: (newBool)
		[`${toFunctionSelector('function setString(string)')}-96`]: 1, // each 32 bytes: (newString (offset), newString (length), newString (value padded to 32 bytes))
		[`${toFunctionSelector('function setAddress(address)')}-32`]: 1, // each 32 bytes: (newAddress)
	}

	// View function call
	const getNumberExpectedTrace = {
		[`${toFunctionSelector('function getNumber()')}-0`]: 1,
	}

	// Call to external contract
	const callMathHelperExpectedTrace = {
		[`${toFunctionSelector('function callMathHelper(uint256)')}-32`]: 1, // each 32 bytes: (newNumber)
		[`${toFunctionSelector('function multiply(uint256)')}-32`]: 1, // each 32 bytes: (newNumber)
	}

	// Delegate call to external contract
	const delegateCallMathHelperExpectedTrace = {
		[`${toFunctionSelector('function delegateCallMathHelper(uint256)')}-32`]: 1, // each 32 bytes: (newNumber)
		[`${toFunctionSelector('function multiply(uint256)')}-32`]: 1, // each 32 bytes: (newNumber)
	}

	// Function call that reverts (should still include the function selector)
	const revertWithStringErrorExpectedTrace = {
		[`${toFunctionSelector('function revertWithStringError()')}-0`]: 1,
	}

	beforeEach(async () => {
		client = createTevmNode()

		const { createdAddress: createdAdvancedContractAddress } = await deployHandler(client)({
			...AdvancedContract.deploy(1n, false, 'test', PREFUNDED_ACCOUNTS[0].address),
			addToBlockchain: true,
		})
		assert(createdAdvancedContractAddress, 'createdAdvancedContractAddress is undefined')
		advancedContractAddress = createAddress(createdAdvancedContractAddress)

		const { createdAddress: createdErrorContractAddress } = await deployHandler(client)({
			...ErrorContract.deploy(),
			addToBlockchain: true,
		})
		assert(createdErrorContractAddress, 'createdErrorContractAddress is undefined')
		errorContractAddress = createAddress(createdErrorContractAddress)
	})

	it('should collect 4-byte function selectors from contract calls', async () => {
		// Call the contract with fourbyteTracer
		const vm = await client.getVm().then((vm) => vm.deepCopy())
		const head = await vm.blockchain.getCanonicalHeadBlock()
		await vm.stateManager.setStateRoot(head.header.stateRoot)

		const params = {
			data: hexToBytes(
				encodeFunctionData(AdvancedContract.write.setAllValues(2n, true, 'test', PREFUNDED_ACCOUNTS[0].address)),
			),
			gasLimit: 16784800n,
			to: advancedContractAddress,
			block: await vm.blockchain.getCanonicalHeadBlock(),
			origin: createAddress(0),
			caller: createAddress(0),
		}

		const result = await runCallWithFourbyteTrace(vm, client.logger, params)
		expect(result.trace).toMatchSnapshot()
		expect(result.trace).toStrictEqual(setAllValuesExpectedTrace)
	})

	it('should handle view function calls', async () => {
		const vm = await client.getVm().then((vm) => vm.deepCopy())
		const head = await vm.blockchain.getCanonicalHeadBlock()
		await vm.stateManager.setStateRoot(head.header.stateRoot)

		const params = {
			data: hexToBytes(encodeFunctionData(AdvancedContract.read.getNumber())),
			gasLimit: 16784800n,
			to: advancedContractAddress,
			block: await vm.blockchain.getCanonicalHeadBlock(),
			origin: createAddress(0),
			caller: createAddress(0),
		}

		const result = await runCallWithFourbyteTrace(vm, client.logger, params)
		expect(result.trace).toStrictEqual(getNumberExpectedTrace)
	})

	it('should handle a call to a function in an external contract', async () => {
		const vm = await client.getVm().then((vm) => vm.deepCopy())
		const head = await vm.blockchain.getCanonicalHeadBlock()
		await vm.stateManager.setStateRoot(head.header.stateRoot)

		const params = {
			data: hexToBytes(encodeFunctionData(AdvancedContract.write.callMathHelper(2n))),
			gasLimit: 16784800n,
			to: advancedContractAddress,
			block: await vm.blockchain.getCanonicalHeadBlock(),
			origin: createAddress(0),
			caller: createAddress(0),
		}

		const result = await runCallWithFourbyteTrace(vm, client.logger, params)
		expect(result.trace).toStrictEqual(callMathHelperExpectedTrace)
	})

	it('should handle delegate calls', async () => {
		const vm = await client.getVm().then((vm) => vm.deepCopy())
		const head = await vm.blockchain.getCanonicalHeadBlock()
		await vm.stateManager.setStateRoot(head.header.stateRoot)

		const params = {
			data: hexToBytes(encodeFunctionData(AdvancedContract.write.delegateCallMathHelper(2n))),
			gasLimit: 16784800n,
			to: advancedContractAddress,
			block: await vm.blockchain.getCanonicalHeadBlock(),
			origin: createAddress(0),
			caller: createAddress(0),
		}

		const result = await runCallWithFourbyteTrace(vm, client.logger, params)
		expect(result.trace).toStrictEqual(delegateCallMathHelperExpectedTrace)
	})

	it('should handle a call to a function that reverts', async () => {
		const vm = await client.getVm().then((vm) => vm.deepCopy())
		const head = await vm.blockchain.getCanonicalHeadBlock()
		await vm.stateManager.setStateRoot(head.header.stateRoot)

		const params = {
			data: hexToBytes(encodeFunctionData(ErrorContract.write.revertWithStringError())),
			gasLimit: 16784800n,
			to: errorContractAddress,
			block: await vm.blockchain.getCanonicalHeadBlock(),
			origin: createAddress(0),
			caller: createAddress(0),
		}

		const result = await runCallWithFourbyteTrace(vm, client.logger, params)
		expect(result.trace).toStrictEqual(revertWithStringErrorExpectedTrace)
	})

	it('should handle CREATE operations by ignoring them', async () => {
		const vm = await client.getVm().then((vm) => vm.deepCopy())
		const head = await vm.blockchain.getCanonicalHeadBlock()
		await vm.stateManager.setStateRoot(head.header.stateRoot)

		// Deploy a new contract during execution
		const deployData = encodeDeployData({
			abi: ErrorContract.abi,
			bytecode: ErrorContract.bytecode,
		})

		const params = {
			data: hexToBytes(deployData),
			gasLimit: 16784800n,
			// No 'to' address means this is a CREATE operation
			block: await vm.blockchain.getCanonicalHeadBlock(),
			origin: createAddress(0),
			caller: createAddress(0),
		}

		const result = await runCallWithFourbyteTrace(vm, client.logger, params)
		// Should be empty since CREATE operations don't have function selectors
		expect(Object.keys(result.trace)).toHaveLength(0)
	})

	it('should handle lazilyRun parameter', async () => {
		const vm = await client.getVm().then((vm) => vm.deepCopy())
		const head = await vm.blockchain.getCanonicalHeadBlock()
		await vm.stateManager.setStateRoot(head.header.stateRoot)

		const params = {
			data: hexToBytes(encodeFunctionData(AdvancedContract.write.callMathHelper(2n))),
			gasLimit: 16784800n,
			to: advancedContractAddress,
			block: await vm.blockchain.getCanonicalHeadBlock(),
			origin: createAddress(0),
			caller: createAddress(0),
		}

		const result = await runCallWithFourbyteTrace(vm, client.logger, params, true)
		// Should be empty since EVM didn't run
		expect(Object.keys(result.trace)).toHaveLength(0)
	})
})
