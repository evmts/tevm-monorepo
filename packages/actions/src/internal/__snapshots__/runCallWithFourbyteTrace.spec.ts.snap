// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runCallWithFourbyteTrace > should collect 4-byte function selectors from contract calls 1`] = `
{
  "0x1e26fd33-32": 1,
  "0x349a0054-192": 1,
  "0x3fb5c1cb-32": 1,
  "0x7fcaf666-96": 1,
  "0xe30081a0-32": 1,
}
`;
